### PR TITLE
Adds User's MFA status when an API key is created

### PR DESCRIPTION
**Pull request by advay26**
Addresses https://github.com/NuGet/Engineering/issues/4179

- Created a new `WasCreatedSecurely` column in the Credentials table
- If a user's current session was Multi Factor Authenticated when creating an API key, the table logs a `True` in the `WasCreatedSecurely` column, and `False` otherwise. All entries before the rollout of this feature will be marked `NULL`.
- Added unit tests (for the `False` case when a user isn't MFA'd, simply adapted an existing test)

### CODE REVIEW

File: src/NuGet.Services.Entities/Credential.cs [hunk 1]
```diff
@@ -84,6 +84,8 @@ public Credential(string type, string value, TimeSpan? expiration)
 
         public CredentialRevocationSource? RevocationSourceKey { get; set; }
 
+        public bool WasCreatedSecurely { get; set; }
```
Comment (joelverhagen) [inline src/NuGet.Services.Entities/Credential.cs:87]: This should be nullable (tri-state): secure, insecure, unknown = null (row was created prior to the app code being deploy).

File: src/NuGetGallery/Controllers/UsersController.cs [hunk 2]
```diff
@@ -1032,6 +1032,7 @@ private async Task<CredentialViewModel> GenerateApiKeyInternal(string descriptio
             var newCredential = _credentialBuilder.CreateApiKey(expiration, out string plaintextApiKey);
             newCredential.Description = description;
             newCredential.Scopes = scopes;
+            newCredential.WasCreatedSecurely = User.WasMultiFactorAuthenticated();
```
Comment (joelverhagen) [inline src/NuGetGallery/Controllers/UsersController.cs:1035]: Does this bug affect this value?
https://github.com/NuGet/Engineering/issues/4181

**Review by joelverhagen — CHANGES_REQUESTED** [review]
Small comments/questions.

**Comment by clairernovotny** [comment]
Can we record the type of MFA for later use--the amr claim value: https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#the-amr-claim

That will tell us the strength of the auth for later analysis and/or enforcement.

File: src/NuGetGallery/Controllers/UsersController.cs (hunk 2 above)
Comment (shishirx34) [reply src/NuGetGallery/Controllers/UsersController.cs:1035]: It should not. That bug is likely that we do not store the MFA for aad account upon creation in DB. The above value is drawn from the claims, which are set correctly, which is what you see on the accounts page for 2FA.

**Comment by shishirx34** [comment]
@clairernovotny - I agree that it would be great to have it, I think for this PR it is a little out of scope. Currently we rely on the ACR claim to know if the MFA was used or not and the AMR claims are not yet plugged into our authentication session.

File: src/NuGet.Services.Entities/Credential.cs (hunk 1 above)
Comment (shishirx34) [reply src/NuGet.Services.Entities/Credential.cs:87]: you might have to drop and recreate the migrations to get the default values updated in the local/dev DB.

**Comment by joelverhagen** [comment]
> @clairernovotny Claire Novotny FTE - I agree that it would be great to have it, I think for this PR it is a little out of scope. Currently we rely on the ACR claim to know if the MFA was used or not and the AMR claims are not yet plugged into our authentication session.

Spoke with @clairernovotny offline. Let's proceed w/o the amr claim value. @clairernovotny feel free to open a NuGetGallery issue tracking this and surface it later when you see it fitting concretely into the rollout plan.

File: src/NuGet.Services.Entities/Credential.cs (hunk 1 above)
Comment (advay26) [reply src/NuGet.Services.Entities/Credential.cs]: I made the field a nullable bool, so now entries that were created before the rollout of this feature will have NULL, while newer entries will have false/true based on their authentication.

**Review by shishirx34 — APPROVED** [review]

**Review by joelverhagen — APPROVED** [review]

**Review by zhhyu — APPROVED** [review]

**Pull request merged**